### PR TITLE
Fourslash cleanup

### PR DESCRIPTION
**Pull request by mhegazy**
In preparation of the ts-server work, we need to have a way to run the language service tests against LS objects using different marshaling/unmarshaling mechanisms. We have been using the shim layer inconsistently. This change hides the details of the marshaling/unmarshaling layer from the fourslash test logic. The Fourslash tests only knows about the LanguageServices interface, a marshaling layer test adapter exposes the same interface and handles marshaling and unmarshaling under the covers.

All tests should be running against the basic "native" language service object. I have also added new tests for the shim layer specifically, basically ensuring that the setup, marshaling and unmarshaling is happening correctly. I have added one test per feature to ensure we exercise that code path.

### CODE REVIEW

File: src/harness/harnessLanguageService.ts [hunk 1]
```diff
@@ -303,105 +189,255 @@ module Harness.LanguageService {
...
+                finalLexState,
+                entries
+            };
         }
+    }
 
-        /** Normalize an array of edits by removing overlapping entries and sorting entries on the minChar position. */
-        private normalizeEdits(edits: ts.TextChange[]): ts.TextChange[] {
-            var result: ts.TextChange[] = [];
+    function unwrappJSONCallResult(result: string): any {
```
Comment (RyanCavanaugh) [inline src/harness/harnessLanguageService.ts]: Spelling ("unwrap~~p~~")

File: src/harness/harnessLanguageService.ts [hunk 2]
```diff
@@ -303,105 +189,255 @@ module Harness.LanguageService {
...
+        }
+        dispose(): void { this.shim.dispose({}); }
+    }
 
-            var temp = mapEdits(edits).sort(function (a, b) {
-                var result = a.edit.span.start - b.edit.span.start;
-                if (result === 0)
-                    result = a.index - b.index;
-                return result;
+    export class ShimLanugageServiceAdaptor implements LanguageServiceAdaptor {
```
Comment (RyanCavanaugh) [inline src/harness/harnessLanguageService.ts]: `Adapter` is the more common spelling

**Comment by RyanCavanaugh** [comment]
:+1: this is great

File: src/harness/fourslash.ts [hunk 3]
```diff
@@ -245,11 +246,9 @@ module FourSlash {
 
     export class TestState {
         // Language service instance
-        public languageServiceShimHost: Harness.LanguageService.TypeScriptLS;
+        private languageServiceAdaptorHost: Harness.LanguageService.LanguageServiceAdaptorHost;
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Adaptor -> Adapter

File: src/harness/harnessLanguageService.ts (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/harness/harnessLanguageService.ts]: And Lanugage is mispeelde two.

File: src/harness/fourslash.ts [hunk 4]
```diff
@@ -275,19 +272,35 @@ module FourSlash {
...
+                this.languageServiceAdaptorHost.addScript(referenceFilePath, inputFile);
             }
         }
 
-        constructor(public testData: FourSlashData) {
-            // Initialize the language service with all the scripts
-            this.cancellationToken = new TestCancellationToken();
-            this.languageServiceShimHost = new Harness.LanguageService.TypeScriptLS(this.cancellationToken);
+        private getLanguageServiceAdaptor(testType: FourSlashTestType):
+            { new (cancellationToken?: ts.CancellationToken, options?: ts.CompilerOptions): Harness.LanguageService.LanguageServiceAdaptor } {
```
Comment (CyrusNajmabadi) [inline src/harness/fourslash.ts]: can we please have an interface for this type.

Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: I'm not sure if this is actually better than `LanguageServiceAdaptorFactory` (which is kind of terrible on its own).

File: src/harness/fourslash.ts [hunk 5]
```diff
@@ -275,19 +272,35 @@ module FourSlash {
...
+            switch (testType) {
+                case FourSlashTestType.Native:
+                    return Harness.LanguageService.NativeLanugageServiceAdaptor;
+                    break;
+                case FourSlashTestType.Shims:
+                    return Harness.LanguageService.ShimLanugageServiceAdaptor;
+                    break;
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Redundant `break`s, but I can live with it.

File: src/harness/fourslash.ts [hunk 6]
```diff
@@ -302,18 +315,16 @@ module FourSlash {
...
-                var importedFiles: ts.IFileReference[] = resolvedResult.importedFiles;
+                var resolvedResult = languageServiceAdaptor.getPreProcessedFileInfo(startResolveFileRef.fileName, startResolveFileRef.content);
+                var referencedFiles: ts.FileReference[] = resolvedResult.referencedFiles;
+                var importedFiles: ts.FileReference[] = resolvedResult.importedFiles;
 
                 // Add triple reference files into language-service host
                 ts.forEach(referencedFiles, referenceFile => {
                     // Fourslash insert tests/cases/fourslash into inputFile.unitName so we will properly append the same base directory to refFile path
-                    var referenceFilePath = "tests/cases/fourslash/" + referenceFile.path;
+                    var referenceFilePath = this.basePath+ '/' + referenceFile.fileName;
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: If we're already relying on node elsewhere in the runner, use `path.join`.

File: src/harness/fourslash.ts [hunk 7]
```diff
@@ -321,29 +332,24 @@ module FourSlash {
                 ts.forEach(importedFiles, importedFile => {
                     // Fourslash insert tests/cases/fourslash into inputFile.unitName and import statement doesn't require ".ts"
                     // so convert them before making appropriate comparison
-                    var importedFilePath = "tests/cases/fourslash/" + importedFile.path + ".ts";
+                    var importedFilePath = this.basePath + '/' + importedFile.fileName + ".ts";
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: `path.join`

File: src/harness/fourslash.ts (hunk 6 above)
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: If not, then space before the `+`.

File: src/harness/fourslash.ts [hunk 8]
```diff
@@ -2320,7 +2272,7 @@ module FourSlash {
                                 currentFileName = fileName;
                             }
 
-                            currentFileName = 'tests/cases/fourslash/' + match[2];
+                            currentFileName = basePath +  '/' + match[2];
```
Comment (yuit) [inline src/harness/fourslash.ts]: Extra space between ```+```

Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Use `join`

File: src/harness/fourslash.ts [hunk 9]
```diff
@@ -453,7 +464,7 @@ module FourSlash {
         private getAllDiagnostics(): ts.Diagnostic[] {
             var diagnostics: ts.Diagnostic[] = [];
 
-            var fileNames = JSON.parse(this.languageServiceShimHost.getScriptFileNames());
+            var fileNames = this.languageServiceAdaptorHost.getFilenames();
```
Comment (yuit) [inline src/harness/fourslash.ts]: it should be name ```getFileNames```

File: src/harness/harnessLanguageService.ts [hunk 10]
```diff
@@ -303,105 +189,255 @@ module Harness.LanguageService {
...
-                            errorString += "                  Left File: " + text1Lines[i] + "\n";
-                            errorString += "                 Right File: " + text2Lines[i] + "\n\n";
-                        }
-                    }
-                    throw (new Error(errorString));
-                }
-            }
-            assert.isTrue(noDiff(formattedScript, baseline));
-            assert.equal(formattedScript, baseline);
+    export interface LanguageServiceAdaptorHost extends LanguageServiceHostBase {
```
Comment (yuit) [inline src/harness/harnessLanguageService.ts]: Why use interface instead of type alias

**Pull request merged**